### PR TITLE
Support basic auth for tidb-server

### PR DESCRIPTION
**Pull request by shenli**
Use password to auth user.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -383,6 +382,71 @@ func (s *session) Close() error {
...
+	if len(strs) != 2 {
+		log.Warnf("Invalid format for user: %s", user)
+		return false
+	}
+	// Get user password.
+	name := strs[0]
+	host := strs[1]
+	authSQL := fmt.Sprintf("SELECT Password FROM %s.%s WHERE User=\"%s\" and Host=\"%s\";", mysql.SystemDB, mysql.UserTable, name, host)
+	rs, err := s.Execute(authSQL)
+	if err != nil {
```
Comment (qiuyesuifeng) [inline session.go]: Why discard this error?

File: tidb-server/server/driver.go [hunk 2]
```diff
@@ -50,6 +50,9 @@ type IContext interface {
 
 	// Close closes the IContext.
 	Close() error
+
+	// Auth auths user
```
Comment (qiuyesuifeng) [inline tidb-server/server/driver.go]: End of .

File: tidb-server/server/conn.go [hunk 3]
```diff
@@ -210,20 +180,27 @@ func (cc *clientConn) readHandshakeResponse() error {
...
+	if err != nil {
+		cc.Close()
+		return errors.Trace(err)
+	}
+	// Do Auth
+	addr := cc.conn.RemoteAddr().String()
+	strs := strings.Split(addr, ":")
+	host := strs[0]
+	user := fmt.Sprintf("%s@%s", cc.user, host)
+	if !cc.server.skipAuth() && !cc.ctx.Auth(user, auth, cc.salt) {
```
Comment (qiuyesuifeng) [inline tidb-server/server/conn.go]: It is better to check `cc.server.skipAuth()` above, like in line 196.

File: tidb-server/server/server_test.go [hunk 4]
```diff
@@ -219,3 +219,18 @@ func runTestConcurrentUpdate(c *C) {
...
+	})
+	newDsn := "test:123@tcp(localhost:4001)/test?strict=true"
+	runTests(c, newDsn, func(dbt *DBTest) {
+		dbt.mustExec(`USE mysql;`)
+	})
+
+	db, err := sql.Open("mysql", "test:456@tcp(localhost:4001)/test?strict=true")
+	_, err = db.Query("USE mysql;")
+	c.Assert(err, NotNil, Commentf("Wrong password should be failed"))
+	defer db.Close()
```
Comment (qiuyesuifeng) [inline tidb-server/server/server_test.go]: Needless to use defer here.

**Comment by shenli** [comment]
PTAL @qiuyesuifeng @coocood

**Pull request merged**